components/script/dom/bluetooth.rs

```diff
@@ -285,31 +292,72 @@ fn canonicalize_filter(filter: &BluetoothRequestDeviceFilter) -> Fallible<Blueto
     };
 
-    // Step 2.4.6.
-    let manufacturer_id = filter.manufacturerId;
-
-    // Step 2.4.7.
-    let service_data_uuid = match filter.serviceDataUUID {
-        Some(ref service_data_uuid) => {
-            // Step 2.4.7.1 - 2.4.7.2.
-            let uuid = try!(BluetoothUUID::service(service_data_uuid.clone())).to_string();
-
-            // Step 2.4.7.3.
-            if uuid_is_blacklisted(uuid.as_ref(), Blacklist::All) {
-                return Err(Security)
+    // Step 2.4.6 - 2.4.7
+    let manufacturer_data = match filter.manufacturerData {
+        Some(ref manufacturer_data_map) => {
+            if manufacturer_data_map.deref().is_empty() {
+                return Err(Type(MANUFACTURER_DATA_ERROR.to_owned()));
             }
+            let mut map = HashMap::new();
+            for (key, bdfi) in manufacturer_data_map.deref() {
+                let manufacturer_id = match u16::from_str(key.as_ref()) {
+                    Ok(id) => id,
+                    Err(err) => return Err(Type(format!("{} {} {}", KEY_CONVERSION_ERROR, key, err))),
+                };
+                map.insert(manufacturer_id, try!(canonicalize_bluetooth_data_filter_init(bdfi)));
+            }
+            Some(map)
+        },
+        None => None,
+    };
 
-            // Step 2.4.7.4.
-            uuid
+    // Step 2.4.8 -2.4.9
+    let service_data = match filter.serviceData {
+        Some(ref service_data_map) => {
+            if service_data_map.deref().is_empty() {
+                return Err(Type(SERVICE_DATA_ERROR.to_owned()));
+            }
+            let mut map = HashMap::new();
+            for (key, bdfi) in service_data_map.deref() {
+                let service_name = match u32::from_str(key.as_ref()) {
+                    Ok(number) => StringOrUnsignedLong::UnsignedLong(number),
+                    _ => StringOrUnsignedLong::String(key.clone())
+                };
+                let service = try!(BluetoothUUID::service(service_name)).to_string();
+                if uuid_is_blacklisted(service.as_ref(), Blacklist::All) {
+                    return Err(Security);
+                }
+                map.insert(service, try!(canonicalize_bluetooth_data_filter_init(bdfi)));
+            }
+            Some(map)
         },
-        None => String::new(),
+        None => None,
     };
 
-    Ok(BluetoothScanfilter::new(name,
-                                name_prefix,
-                                services_vec,
-                                manufacturer_id,
-                                service_data_uuid))
+    // Step 10.
+    Ok(BluetoothScanfilter::new(name, name_prefix, services_vec, manufacturer_data, service_data))
+}
+
+// https://webbluetoothcg.github.io/web-bluetooth/#bluetoothdatafilterinit-canonicalizing
+fn canonicalize_bluetooth_data_filter_init(bdfi: &BluetoothDataFilterInit) -> Fallible<(Vec<u8>, Vec<u8>)> {
+    // Step 1.
+    let data_prefix = match bdfi.dataPrefix {
```

Comment:
Note: I changed this to `clone` not `cloned`. I'm not sure why was the `ref` used here.